File: instrumentation/spray-http-1.3.1/src/main/scala/com/agent/instrumentation/spray/ResponseWrapper.java
/*
 *
 *  * Copyright 2020 New Relic Corporation. All rights reserved.
 *  * SPDX-License-Identifier: Apache-2.0
 *
 */

package com.agent.instrumentation.spray;

import java.util.List;

import com.newrelic.api.agent.HeaderType;
import com.newrelic.api.agent.Response;
import spray.http.HttpHeader;
import spray.http.HttpHeaders;
import spray.http.StatusCode;

public class ResponseWrapper implements Response {

    private final StatusCode status;
    private final List<HttpHeader> headers;

    public ResponseWrapper(StatusCode status, List<HttpHeader> headers) {
        this.status = status;
        this.headers = headers;
    }

    @Override
    public int getStatus() throws Exception {
        return status.intValue();
    }

    @Override
    public String getStatusMessage() throws Exception {
        return status.reason();
    }

    @Override
    public String getContentType() {
        for (HttpHeader header : headers) {
            if (header.is("content-type")) {
                return header.value();
            }
        }
        return null;
    }

    @Override
    public HeaderType getHeaderType() {
        return HeaderType.HTTP;
    }

    @Override
    public void setHeader(String name, String value) {
        headers.add(new HttpHeaders.RawHeader(name, value));
    }

}

File: instrumentation/apache-pekko-http-2.13_1/src/main/scala/com/agent/instrumentation/org.apache.pekko/http/ResponseWrapper.scala
/*
 *
 *  * Copyright 2024 New Relic Corporation. All rights reserved.
 *  * SPDX-License-Identifier: Apache-2.0
 *
 */

package com.agent.instrumentation.org.apache.pekko.http

import org.apache.pekko.http.scaladsl.model.HttpResponse
import org.apache.pekko.http.scaladsl.model.headers.RawHeader
import com.newrelic.api.agent.{ExtendedResponse, HeaderType}

class ResponseWrapper(var response: HttpResponse) extends ExtendedResponse {

  def getStatus: Int = {
    response.status.intValue
  }

  def getStatusMessage: String = {
    response.status.reason
  }

  def getContentType: String = {
    response.entity.contentType.value
  }

  def getHeaderType: HeaderType = {
    HeaderType.HTTP
  }

  def setHeader(name: String, value: String): Unit = {
    response = response.addHeader(new RawHeader(name, value))
  }

  def getContentLength: Long = {
    val contentLength = response.getHeader("Content-Length")
    if (contentLength.isPresent) {
      return contentLength.get().value().toLong
    }
    response.entity.getContentLengthOption().orElse(-1L)
  }

}
